### PR TITLE
[AArch64] Drop poison-generating flags in `genSubAdd2SubSub` combiner

### LINKED ISSUE #88950: AArch64 backend miscompilation related to sshl.sat
here's a function:
```llvm
declare i64 @llvm.sshl.sat.i64(i64, i64) #0

define i64 @f(i64 %0, i64 %1, i64 %2) {
  %4 = or i64 %0, 1
  %5 = add i64 %4, %0
  %6 = sub nsw i64 %1, %5
  %7 = call i64 @llvm.sshl.sat.i64(i64 %6, i64 1)
  ret i64 %7
}

attributes #0 = { nocallback nofree nosync nounwind speculatable willreturn memory(none) }
```
let's invoke it using this driver:
```c
#include <stdio.h>

// i64 %#0 = #x5dbdff5f99a88000 (6754835802190544896)
// i64 %#1 = #x0000000000000000 (0)
// i64 %#2 = #x0000000000000000 (0)

unsigned long f(unsigned long, unsigned long, unsigned long);

int main(void) {
  printf("%lx\n", f(6754835802190544896ULL, 0, 0));
}
```
the result is:
```
Johns-MacBook-Pro:reduce regehr$ ~/llvm-project/for-alive/bin/llc reduced.ll
Johns-MacBook-Pro:reduce regehr$ clang driver.c reduced.s
Johns-MacBook-Pro:reduce regehr$ ./a.out 
8000000000000000
Johns-MacBook-Pro:reduce regehr$ 
```
but it looks like the answer should be LONG_MAX instead. here I'll just give Alive's work, which appears correct:
```
i64 %#4 = #x5dbdff5f99a88001 (6754835802190544897)
i64 %#5 = #xbb7bfebf33510001 (13509671604381089793, -4937072469328461823)
i64 %#6 = #x44840140ccaeffff (4937072469328461823)
i64 %#7 = #x7fffffffffffffff (9223372036854775807)
```

### PR DESCRIPTION
**Pull request by antoniofrighetto**
A miscompilation issue has been addressed with improved handling.

Fixes: https://github.com/llvm/llvm-project/issues/88950.

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 1]
```diff
@@ -6924,14 +6924,22 @@ genSubAdd2SubSub(MachineFunction &MF, MachineRegisterInfo &MRI,
     assert((Opcode == AArch64::SUBWrr || Opcode == AArch64::SUBXrr) &&
            "Unexpected instruction opcode.");
 
+  uint32_t RootFlags = Root.getFlags();
+  if (RootFlags & MachineInstr::NoSWrap)
+    Root.clearFlag(MachineInstr::NoSWrap);
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:6929]: I think this can just call clearFlag, and it will clear it if it is present.

**Review by davemgreen — COMMENTED** [review]
Thanks for putting the patch together.

I think the flags were being added by https://github.com/llvm/llvm-project/blob/fe47e8ff3ae7fc8975eaade6bfa6679737c28b93/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp#L7880? If this patch clears the flags on root then they will no longer be set on the new instructions, which is what we want.

It might be worth having the switch at https://github.com/llvm/llvm-project/blob/fe47e8ff3ae7fc8975eaade6bfa6679737c28b93/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp#L6969 return as opposed to break, so that we don't run the code that is intended for mul+add. Either way this looks correct now, but it would mean that this code doesn't need to modify Root.

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 2]
```diff
@@ -6924,14 +6924,22 @@ genSubAdd2SubSub(MachineFunction &MF, MachineRegisterInfo &MRI,
     assert((Opcode == AArch64::SUBWrr || Opcode == AArch64::SUBXrr) &&
            "Unexpected instruction opcode.");
 
+  uint32_t RootFlags = Root.getFlags();
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:6927]: Perhaps use `Root.mergeFlagsWith(*AddMI);`, clear the NoSWrap/NoUWrap on it they use that in the new instructions?

**Comment by antoniofrighetto** [comment]
Thanks for reviewing it!

> It might be worth having the switch at
> 
> https://github.com/llvm/llvm-project/blob/fe47e8ff3ae7fc8975eaade6bfa6679737c28b93/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp#L6969
> 
> return as opposed to break, so that we don't run the code that is intended for mul+add. Either way this looks correct now, but it would mean that this code doesn't need to modify Root.

Right, although clearing the flags makes explicit the semantic of dropping flags in the combiner. I'd personally leave the `break`, as we would need to anticipate the following in both SUBADD_OP1/SUBADD_OP2 cases otherwise, no strong concerns though.

https://github.com/llvm/llvm-project/blob/fe47e8ff3ae7fc8975eaade6bfa6679737c28b93/llvm/lib/Target/AArch64/AArch64InstrInfo.cpp#L7876

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 3]
```diff
@@ -6924,14 +6924,20 @@ genSubAdd2SubSub(MachineFunction &MF, MachineRegisterInfo &MRI,
     assert((Opcode == AArch64::SUBWrr || Opcode == AArch64::SUBXrr) &&
            "Unexpected instruction opcode.");
 
+  Root.mergeFlagsWith(*AddMI);
```
Comment (v01dXYZ) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:6927]: Hi,
I think you made a little mistake as `mergeFlagsWith` is a const method.

**Comment by v01dXYZ** [comment]
@davemgreen Out of curiosity, you seem to prefer not to not mutate `Root` if possible.

Is `Root` always replaced ? (it doesn't seem so but I prefer to ask).

File: llvm/lib/Target/AArch64/AArch64InstrInfo.cpp [hunk 4]
```diff
@@ -6924,14 +6924,20 @@ genSubAdd2SubSub(MachineFunction &MF, MachineRegisterInfo &MRI,
     assert((Opcode == AArch64::SUBWrr || Opcode == AArch64::SUBXrr) &&
            "Unexpected instruction opcode.");
 
+  uint32_t Flags = Root.mergeFlagsWith(*AddMI);
+  Root.clearFlag(MachineInstr::NoSWrap);
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrInfo.cpp:6928]: Flags = Flags & ~MachineInstr::NoSWrap;
Flags = Flags & ~MachineInstr::NoUWrap;

**Review by davemgreen — COMMENTED** [review]
Hi - yeah I was expecting Root to always be replaced, hence it should be correct either way. Not modifying Root feels a little cleaner though. You would not usually expect the old code to be modified as the new nodes are created, and it can be a bit confusing to follow the logic of setting the on the Root, to be used later on (in code that is easy to miss), to override the flags of the instructions that have just been created.

> DelInstrs.push_back(&Root); 

Yeah I was thinking that would need to be added, either to the switch or into the genSubAdd2SubSub method.

**Comment by antoniofrighetto** [comment]
Should be on track now!

**Review by davemgreen — APPROVED** [review]
Thanks. LGTM

**Comment by momchil-velikov** [comment]
It looks like there's a slight issue with the test. If you have enabled expensive checks, the combiner calls `verifyPatternOrder` which creates(and discards) a few instructions, but in doing so consumes some virtual register numbers. Then you get unexpected output in the test, for example, I get `%9` instead of `%7` in these check lines:
```
# CHECK-LABEL: name: test8
# CHECK:       %7:gpr64 = SUBXrr %1, %0
# CHECK-NEXT:  %4:gpr64common = SUBXrr killed %7, killed %2
```

**Comment by antoniofrighetto** [comment]
@momchil-velikov Thank you for reporting, and sorry to get back only now, somehow I missed the notification for this. Fixed in ca257022aa130f95b2530f9aba98af8d63027b0d.

**Pull request merged**